src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs

```diff
@@ -168,28 +169,39 @@ private bool TryConvertToXml(string xml, out object doc, ref Exception exRef)
 
         private bool TryConvertToJson(string json, out object obj, ref Exception exRef)
         {
+            bool converted;
             try
             {
                 ErrorRecord error;
                 obj = JsonObject.ConvertFromJson(json, out error);
+                converted = true;
```

Comment:
I wonder why the fix so high level? It seems `ConvertFromJson` or just `PopulateFromJDictionary` is right place to fix.
We should think to return `System.Management.Automation.Internal.AutomationNull`.

Also please review [JsonSerializerSettings](https://www.newtonsoft.com/json/help/html/NullValueHandlingIgnore.htm).

Reply:
@iSazonov `ConvertFromJson()` was already returning null. The problem is that `TryConvertToJson()` was using null to mean that JSON serialization failed. We want to return null and idealy keep what ever `ConvertFromJson()` was already returning. But, `ConvertFromJson()` can return null in situations that don't mean it was JSON `null` (at least I'm pretty sure it does). In that case we should let Invoke-RestMethod continue with i's other serialization tests by returning false. This has to be fixed here because it's a problem with Invoke-RestMethod's logic, not with `ConvertFromJson()`

Reply:
Hmm i realize this is probably very confusing. The an API endpoint returns the following:

```none
HTTP/1.1 200 OK
Content-Length: 4
Content-Type: application/json
Date: Sat, 04 Nov 2017 18:49:54 GMT

null
```
that `null` is the JSON literal for `$null`. `ConvertFromJson()` properly returns `$null` for that, but `Invoke-RestMethod` was interpreting that `$null` as failing to serialize the object. `TryConvertToJson()` was returning false and `Invoke-RestMethod` would move on to other serialization methods until ultimately falling back to a string and returning the string `null`.

`ConvertFromJson()` returns null  for empty strings too, however, in this instance, an empty or whitespace string should be returned as an an empty or whitespace string, not null, because an empty or white space string is not valid JSON.  I guess it could be argued that maybe that behavior of `ConvertFromJson()` should be changed and that it should throw on an empty or whitespace string. But, I don't see that going over well.

In any case, because `ConvertFromJson()` behaves like that I'm double checking it is a true JSON `null`.

Reply:
We have the [workarround](https://github.com/PowerShell/PowerShell/blob/4bc52d2358222084738157a08907fac32d13bd3a/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L71). We can remove it - the bug was fixed in Newton.Json 10.0.3. But I think we can keep `JArray.Parse(input);` to throw on any error in Json input string.

Reply:
@iSazonov this has nothing to do with arrays and that code path is not hit in the instace this is addressing. This is about when `null` is NOT in a JSON array or object, but is solitary, by it self, a single value and the only value in the JSON string. this is literally the JSON string this PR addresses:

```json
null
```

This has nothing to do with either of these which already work properly:

```json
[
  null
]
````
```json
{
  "key": null
}
```

Removing that workaround in `ConvertFromJson()` will not affect the logic in this PR.

Reply:
What about `JToken.Parse()`? I believe it throw on null string.

Reply:
Are you suggesting to use `JToken.Parse()` instead of the regex? I guess we could do that. It does throw on empty and whitespace only strings.

`ConvertFromJson()` does both. It throws and sets the `out Error`. Invalid JSON throws, trouble serializing JSON into PowerShell objects gives errors.

Reply:
Yes, sorry I walk around :-), I think we should use `JToken.Parse()` - maybe it'll catch other mistakes we don't know yet.

As for `ConvertFromJson()` - I guess it is common convention - no throw if there is `out Error`?

Reply:
> As for `ConvertFromJson()` - I guess it is common convention - no throw if there is out Error?

I'm not sure what you are asking.

Here are the kinds of strings it throws on: `{`, `[`, `[ null null]`
Here is a string it errors on: `{ "": "value"}`

Errors arise from valid JSON that cannot serialized into PowerShell objects while exceptions arise from invalid JSON. The only invalid JSON I'm aware of that doesn't throw an exception is `String.Empty` and a string with just whitespace. An empty string and whitespace-only string are technically not valid JSON.

In any case, when there is an error it returns null. so we can just ensure that the null it is returning is an actual `null` literal with `JToken.Parse()` In `Invoke-RestMethod`.  It might be a bit redundant, but I will run it on a null check only. Otherwise, I'd rather trust everything else that comes from `ConvertFromJson()`. It's just that the existing `Invoke-RestMethod` logic was assuming null was always a failure, when null is a legitimate single value for JSON.

But, this does bring to light some shortcomings of `ConvertFromJson()`. I think you are the better person to decide how serious they are. but if we want/need to change that behavior, we need to do it before 6.0.0 as I doubt we can get away with that kind of breaking change until 7.0.0. What do you think?

Reply:
I really don't like the API but we already have no time to discuss and make changes in depth. 😕 

I'm sure we'll be with CoreFX implementation in 7.0. 😄 So I would document the API as obsolete or make it `internal` - the later allow us to migrate on CoreFX implementation in any time before 7.0.